### PR TITLE
refactor(batch): Use futures-async-stream to implement HashJoin executor 

### PR DESCRIPTION
**Pull request by D2Lark**
## What's changed and what's your intention?

Implement HashJoin using Executor2 trait.

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
close https://github.com/singularity-data/risingwave/issues/1947

### CODE REVIEW

**Comment by D2Lark** [comment]
There are still some unresolved issues, such as Calling child.execute() twice at the same time shows the problem of ownership

>    --> src/batch/src/executor2/join/hash_join.rs:174:17
>     |
> 170 |         let mut right_child_stream = self.right_child.execute();
>     |                                                       --------- `self.right_child` partially moved due to this method call
> ...
> 174 |                 self.build(build_table, right_child_stream).await?
>     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ value borrowed here after partial move
>     |
> note: this function takes ownership of the receiver `self`, which moves `self.right_child`
>    --> src/batch/src/executor2/mod.rs:67:16
>     |
> 67  |     fn execute(self: Box<Self>) -> BoxedDataChunkStream;
>     |                ^^^^
>     = note: partial move occurs because `self.right_child` has type `std::boxed::Box<dyn executor2::Executor2>`, which does not implement the `Copy` trait
> 
> error[E0382]: use of moved value: `left_child_stream`

File: src/batch/src/executor2/join/chunked_data.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
+// Copyright 2022 Singularity Data
```
Comment (liurenjie1024) [inline src/batch/src/executor2/join/chunked_data.rs:1]: We should also remove `chunked_data` in `executor/join` mod?

File: src/batch/src/executor2/join/hash_join.rs [hunk 2]
```diff
@@ -94,11 +97,11 @@ impl<K> Default for HashJoinState<K> {
...
 
-pub(super) struct HashJoinExecutor<K> {
+pub(super) struct HashJoinExecutor2<K> {
     /// Probe side
-    left_child: BoxedExecutor,
+    left_child: BoxedExecutor2,
     /// Build side
-    right_child: BoxedExecutor,
+    right_child: BoxedExecutor2,
     state: HashJoinState<K>,
```
Comment (liurenjie1024) [inline src/batch/src/executor2/join/hash_join.rs:105]: In fact, after using stream api, we no longer need this state machine. We just need to  keep `BuildTable` and `ProbeTable<K>` as local varialbe in `do_execute` method.

**Comment by liurenjie1024** [comment]
> 

Yes, `execute` will take the ownership of executor.

**Comment by D2Lark** [comment]
> Yes, `execute` will take the ownership of the executor.
yes, because of that we usually put the `execute` at the end of func. So far there are no problems in the non-join executor. But every join executor has two children to take their father's ownership , What to do with this？

File: src/batch/src/executor2/join/hash_join.rs (hunk 2 above)
Comment (D2Lark) [reply src/batch/src/executor2/join/hash_join.rs:105]: From my simple understanding, the previous control flow `open next close `caused a lot of variables to be bound to the executor and the whole logic was very complicated. So this morning I made a [version o](https://github.com/singularity-data/risingwave/pull/2102)f the risinglight implementation and simplified it

Comment (liurenjie1024) [reply src/batch/src/executor2/join/hash_join.rs:105]: Thanks for you contribution, I'll take a review.

**Comment by skyzh** [comment]
> > Yes, `execute` will take the ownership of the executor.
> 
> yes, because of that we usually put the `execute` at the end of func. So far there are no problems in the non-join executor. But every join executor has two children to take their father's ownership , What to do with this？

Just set them as `Option<BoxedExecutor>` and use `left.take().unwrap()` to get the ownership.

File: src/batch/src/executor2/join/chunked_data.rs (hunk 1 above)
Comment (D2Lark) [reply src/batch/src/executor2/join/chunked_data.rs:1]: I tried this once but had some problems, if I delete` executor/join/chunked_data`, many functions and structures under `executor2/join/chunked_data` have to be changed to pub, so I keep `executor/join/chunked_data` for now and delete it after all of them have been ported?

**Comment by D2Lark** [comment]
> > > Yes, `execute` will take the ownership of the executor.
> > 
> > 
> > yes, because of that we usually put the `execute` at the end of func. So far there are no problems in the non-join executor. But every join executor has two children to take their father's ownership , What to do with this？
> 
> Just set them as `Option<BoxedExecutor>` and use `left.take().unwrap()` to get the ownership.

Thanks for the tip~

File: src/batch/src/executor2/join/hash_join.rs (hunk 2 above)
Comment (D2Lark) [reply src/batch/src/executor2/join/hash_join.rs:105]: > `BuildTable` and `ProbeTable<K>` a

I see what you mean, I'll optimize it

**Comment by TennyZhuang** [comment]
CI running https://github.com/singularity-data/risingwave/actions/runs/2237323815

**Comment by TennyZhuang** [comment]
rustdoc check failed @D2Lark 

```
error: unresolved link to `super::hash_join::HashJoinExecutor::probe_remaining`
[62](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:62)
   --> src/batch/src/executor/join/nested_loop_join.rs:350:22
[63](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:63)
    |
[64](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:64)
350 |     /// Similar to [`super::hash_join::HashJoinExecutor::probe_remaining`]. For nested loop join,
[65](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:65)
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no item named `hash_join` in module `join`
[66](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:66)
    |
[67](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:67)
note: the lint level is defined here
[68](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:68)
   --> src/batch/src/lib.rs:26:9
[69](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:69)
    |
[70](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:70)
26  | #![deny(rustdoc::broken_intra_doc_links)]
[71](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:71)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[72](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:72)

[73](https://github.com/singularity-data/risingwave/runs/6205527285?check_suite_focus=true#step:10:73)
error: could not document `risingwave_batch`
```

**Review by liurenjie1024 — APPROVED** [review]
LGTM

**Comment by TennyZhuang** [comment]
https://github.com/singularity-data/risingwave/actions/runs/2238380756 CI running

**Comment by yuhao-su** [comment]
LGTM

**Review by yuhao-su — APPROVED** [review]

**Comment by TennyZhuang** [comment]
https://github.com/singularity-data/risingwave/actions/runs/2238638430

**Comment by liurenjie1024** [comment]
> https://github.com/singularity-data/risingwave/actions/runs/2238638430

The check passed, merged.

**Pull request merged**